components/script/dom/htmlinputelement.rs

```diff
@@ -1034,8 +1034,19 @@ impl HTMLInputElement {
             InputType::Time => {
                 let mut textinput = self.textinput.borrow_mut();
 
-                if ! textinput.single_line_content().is_valid_time_string() {
-                    *textinput.single_line_content_mut() = "".into();
+                if !textinput.single_line_content().is_valid_time_string() {
+                    textinput.single_line_content_mut().clear();
+                }
+            }
+            InputType::DatetimeLocal => {
+                let mut textinput = self.textinput.borrow_mut();
+                if textinput.single_line_content().is_valid_local_date_and_time_string() {
+                    if textinput.single_line_content_mut()
```

Comment:
Is it possible to remove the check for valid string and still produce the correct behavior?

Reply:
I don't think so. SPEC says check valid first, and if valid make it normalized.
It should be two functions to do that.

Reply:
The point here is, since the parsing is already doing the validation anyway, if the parse function is returning an error, then it means it is not a valid string.

Reply:
It's true. Both `is_valid_local_date_and_time_string` and `convert_valid_normalized_local_date_and_time_string` call `parse_local_date_and_time_string`.  I don't combine them because I am afraid that each unit function would be used in other place and in this point, combining them is not a good choice.

Reply:
I do not mean combining them, I meant whether it was possible to just use `convert_valid_normalized_local_date_and_time_string` instead of calling both `is_valid_local_date_and_time_string` and `convert_valid_normalized_local_date_and_time_string`. If so, then we should just use `convert_valid_normalized_local_date_and_time_string`.

Reply:
It's possible.